### DISCUSSION TITLE
Ignore type errors

### DISCUSSION THREAD

**Discussion by jprochazk** [nlohmann/json#2400] [Q&A] [answered]
I'm trying to deserialize unreliable data from a network endpoint. I can overload `adl_serializer` to allow for serializing to/from `std::optional`, and I can set `allow_exceptions` to false in `json::parse` to discard values instead of throwing syntax/parse errors, but I can't find a way to discard values instead of throwing type errors, without using try/catch.

This following will print `Caught exception: [json.exception.type_error.302] type must be number, but is string`. Is there a way to ignore the type error (without try/catch), and return a discarded value instead?

```c++
#include <cstdint>
#include <iostream>
#include <nlohmann/json.hpp>
#include <optional>
using json = nlohmann::json;

namespace nlohmann {
template<typename T> struct adl_serializer<std::optional<T>> {
    static void to_json(json& j, const std::optional<T>& opt) {
        if (!opt.has_value()) {
            j = nullptr;
        } else {
            j = *opt;
        }
    }

    static void from_json(const json& j, std::optional<T>& opt) {
        if (j.is_null()) {
            opt = {};
        } else {
            opt = j.get<T>();
        }
    }
};
}

struct packet {
    std::optional<uint32_t> id;
};

packet receive(const json& data) {
    packet pkt{};
    pkt.id = data["id"].get<std::optional<uint32_t>>();
    return pkt;
}

int main() {
    std::string data = "{ \"id\": \"not a number\" }";
    try {
        receive(json::parse(data, nullptr, false));
    } catch (std::exception& e) {
        std::cerr << "Caught exception: " << e.what() << std::endl;
    }
    return 0;
}
```

**Comment by nlohmann** [comment, marked as answer]
You could use [`is_number()`](https://nlohmann.github.io/json/api/basic_json/is_number/) or [`type()`](https://nlohmann.github.io/json/api/basic_json/type/) to check the type.

**Reply by nlohmann** [reply]
Do you need further assistance with this issue?
